feat: add source code analyzer

tassiLuca opened the pull request:
# Summary
This PR adds a source code tokenizer for Java sources.

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/Analyzer.kt:
```diff
@@ -0,0 +1,18 @@
+ * @param <O> the type of output
+ * @param <T> the type of the representation
+ */
+interface Analyzer<I, out O : SourceRepresentation<T>, T> {
```
This might maybe just extend a normal function:

interface Analyzer<in I, out O : SourceRepresentation<T>, T> : (I) -> O

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/StepHandler.kt:
```diff
@@ -0,0 +1,14 @@
+ * @param I the type of the input.
+ * @param O the type of the output.
+ */
+fun interface StepHandler<I, O> {
```
```suggestion
fun interface StepHandler<in I, out O> : (I) -> O {
```

this looks like it can extend a function.
It actually looks exactly like a function, and since there is no boundary at all on the types, it provides no information but "it's a function from I to O".

maybe it is the case to use just a `typealias StepHandler<I, O> = (I) -> O`?

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/TokenizedSourceImpl.kt:
```diff
@@ -0,0 +1,39 @@
+ */
+data class TokenizedSourceImpl(
+    override val sourceFile: File,
+    override val representation: Sequence<Token>
```
```suggestion
    override val representation: Sequence<Token>,
```

trailing comma

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/TokenizedSourceImpl.kt:
```diff
@@ -0,0 +1,39 @@
+            .map { GramImpl(it.asSequence()) }
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) {
+            return true
+        } else if (javaClass != other?.javaClass) {
+            return false
+        }
+        other as TokenizedSourceImpl
+        return (sourceFile == other.sourceFile) && (representation.toList() == other.representation.toList())
+    }
+
+    override fun hashCode(): Int {
+        var result = sourceFile.hashCode()
+        result = 31 * result + representation.hashCode()
+        return result
```
```suggestion
```

`data class`es come with pre-defined `equals` and `hashCode` that work as they should. Do not override.

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/token/GramImpl.kt:
```diff
@@ -0,0 +1,24 @@
+/**
+ * A simple class implementing a [Gram].
+ */
+data class GramImpl<T>(override val items: Sequence<T>) : Gram<T> {
+    override fun hashCode(): Int {
+        return items.hashCode()
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) {
+            return true
+        } else if (javaClass != other?.javaClass) {
+            return false
+        }
+        other as GramImpl<*>
+        return other.items.toList() == items.toList()
+    }
+
+    override fun toString(): String {
+        return items.toList().toString()
+    }
+}
```
```suggestion
data class GramImpl<T>(override val items: Sequence<T>) : Gram<T> {
    override fun toString(): String = "Gram$items"
}
```

much simpler :)

tassiLuca replied:
During the development, however, [this](https://github.com/tassiLuca/code-plagiarism-detector/blob/master/src/test/kotlin/org/danilopianini/plagiarismdetector/analyzer/TokenizedSourceTest.kt) test failed as you can see [here](https://github.com/tassiLuca/code-plagiarism-detector/actions/runs/3072878148/jobs/4964648487#step:3:271) (line 259). This is due to the fact in the constructor `items` is typed as a `Sequence`. So I override `equals` and `hashCode`. Would it be better to change it in an `Iterable`?

DanySK replied:
If it is an ordered collection of items, the data structure to use in Kotlin is `List`. `Sequence`s are meant as "pull" pipelines (similar to Java's `Stream`s), to be used when the element count is unknown and the processing should be depth-first. `Flow`s are "push" pipelines.

You can still provide a secondary constructor with a sequence, if needed.

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/token/LanguageTokenTypes.kt:
```diff
@@ -0,0 +1,16 @@
+    /**
+     * @return the [TokenType] associated to the given construct.
+     */
+    fun getTokenTypeBy(constructName: String): TokenType
```
shorter names sound more idiomatic
```suggestion
    fun tokenFor(constructName: String): TokenType
```

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/token/LanguageTokenTypes.kt:
```diff
@@ -0,0 +1,16 @@
+    /**
+     * @return true if the given construct name is valid, false otherwise.
+     */
+    fun isValidToken(constructName: String): Boolean
```
```suggestion
    fun isToken(constructName: String): Boolean
```

DanySK reviewed (changes requested):
Reviewed 10/31

tassiLuca commented:
@DanySK in order to increase the code coverage I added a [test](https://github.com/tassiLuca/code-plagiarism-detector/blob/99e3817d2e918cb492bc1ccbb97d8332df1ab42c/src/test/kotlin/org/danilopianini/plagiarismdetector/provider/EnvironmentTokenSupplierTest.kt#L27) which needs the following argument to be added to the JVM that runs the test:
```
--add-opens=java.base/java.util=ALL-UNNAMED
```

I created a [`gradle.properties`](https://github.com/tassiLuca/code-plagiarism-detector/blob/99e3817d2e918cb492bc1ccbb97d8332df1ab42c/gradle.properties#L1) file in which I added the argument and passed it to the [`test` task](https://github.com/tassiLuca/code-plagiarism-detector/blob/99e3817d2e918cb492bc1ccbb97d8332df1ab42c/build.gradle.kts#L64).

Is this the best way to do that?

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/TokenizedSourceImpl.kt:
```diff
@@ -0,0 +1,26 @@
+ */
+data class TokenizedSourceImpl(
+    override val sourceFile: File,
+    val tokens: List<Token>
```
```suggestion
    val tokens: List<Token>,
```

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/TokenizedSourceImpl.kt:
```diff
@@ -0,0 +1,26 @@
+) : TokenizedSource {
+    companion object {
+        private const val SLIDING_STEP = 1
+    }
+    override val representation: Sequence<Token> = tokens.asSequence()
```
```suggestion
    }

    override val representation: Sequence<Token> = tokens.asSequence()
```

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/TokenizedSourceImpl.kt:
```diff
@@ -0,0 +1,26 @@
+    }
+    override val representation: Sequence<Token> = tokens.asSequence()
+
+    override fun splitInGramsOf(size: Int): Sequence<Gram<Token>> {
+        return representation.windowed(size, SLIDING_STEP)
+            .map(::GramImpl)
+    }
```
```suggestion
    override fun splitInGramsOf(size: Int): Sequence<Gram<Token>> = representation
        .windowed(size, SLIDING_STEP)
        .map(::GramImpl)
```

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/token/LanguageTokenTypesImpl.kt:
```diff
@@ -0,0 +1,15 @@
+ * A simple class implementing a [LanguageTokenTypes].
+ * @param tokensTypes the set of token types of the language
+ */
+class LanguageTokenTypesImpl(private val tokensTypes: Collection<TokenType>) : LanguageTokenTypes {
+    override fun tokenFor(constructName: String): TokenType {
```
```suggestion
class LanguageTokenTypesImpl(private val tokensTypes: Collection<TokenType>) : LanguageTokenTypes {

    override fun tokenFor(constructName: String): TokenType {
```

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/token/LanguageTokenTypesImpl.kt:
```diff
@@ -0,0 +1,15 @@
+class LanguageTokenTypesImpl(private val tokensTypes: Collection<TokenType>) : LanguageTokenTypes {
+    override fun tokenFor(constructName: String): TokenType {
+        require(isToken(constructName))
+        return tokensTypes.find { it.languageConstructs.contains(constructName) }!!
```
Why the non-null assertion? This is NPE-prone.

tassiLuca replied:
```kotlin
override fun tokenFor(constructName: String): TokenType =
    tokensTypes.find { constructName in it.languageConstructs } ?: error("No token found for given construct name.")
```

Better?

DanySK replied:
A lot. And even better with a cleaner error:
`error("Token $constructName does not exist. Available constructs: $tokensTypes")`

Failing clearly is very helpful.

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/token/LanguageTokenTypesImpl.kt:
```diff
@@ -0,0 +1,15 @@
+    }
+
+    override fun isToken(constructName: String): Boolean =
+        tokensTypes.any { it.languageConstructs.contains(constructName) }
```
```suggestion
        tokensTypes.any { constructName in it.languageConstructs }
```

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/token/TokenImpl.kt:
```diff
@@ -0,0 +1,12 @@
+data class TokenImpl(
+    override val line: Int,
+    override val column: Int,
+    override val type: TokenType
```
```suggestion
    override val type: TokenType,
```

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/representation/token/TokenTypeImpl.kt:
```diff
@@ -0,0 +1,14 @@
+@Serializable
+data class TokenTypeImpl(
+    override val name: String,
+    override val languageConstructs: Collection<String>
```
```suggestion
    override val languageConstructs: Iterable<String>,
```

tassiLuca replied:
I used a collection because I load the token types from a yaml configuration file.

DanySK replied:
Question: can the constructs appear twice, or would that be an error?

tassiLuca replied:
Would be an error... better a Set.

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/technique/tokenization/java/JavaPreprocessor.kt:
```diff
@@ -0,0 +1,18 @@
+ * accuracy of the system.
+ */
+class JavaPreprocessor : StepHandler<CompilationUnit, CompilationUnit> {
+    override fun invoke(input: CompilationUnit): CompilationUnit {
+        input.removePackageDeclaration()
+        input.findAll(ImportDeclaration::class.java).forEach(input::remove)
+        return input
+    }
```
Prefer immutability, do not modify the input. Make `CompilationUnit` immutable. I want to prevent components to fire back at us when we'll parallelize

tassiLuca replied:
`CompilationUnit` is a third library class. I can clone the input and work on the cloned object.

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/technique/tokenization/java/JavaParser.kt:
```diff
@@ -0,0 +1,18 @@
+ * A java source file parser.
+ */
+class JavaParser : StepHandler<File, CompilationUnit> {
+    override fun invoke(input: File): CompilationUnit {
+        val parserConfiguration = ParserConfiguration().setAttributeComments(false)
+        StaticJavaParser.setConfiguration(parserConfiguration)
+        return StaticJavaParser.parse(input)
+    }
```
```suggestion
    override fun invoke(input: File): CompilationUnit = StaticJavaParser.run {
        setConfiguration(ParserConfiguration().setAttributeComments(false))
        parse(input)
    }
```

DanySK commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/technique/tokenization/java/JavaParser.kt:
```diff
@@ -0,0 +1,18 @@
+class JavaParser : StepHandler<File, CompilationUnit> {
+    override fun invoke(input: File): CompilationUnit {
+        val parserConfiguration = ParserConfiguration().setAttributeComments(false)
+        StaticJavaParser.setConfiguration(parserConfiguration)
```
It looks like you have a mutable shared `object`. This is extremely fragile and should not happen.

DanySK reviewed (changes requested):
Reviewed 20/33

Avoid mutability as much as possible, especially within `object`s.

tassiLuca commented:
@DanySK I finished making the requested changes :)

DanySK commented on src/test/kotlin/org/danilopianini/plagiarismdetector/provider/EnvironmentTokenSupplierTest.kt:
```diff
@@ -0,0 +1,37 @@
+            ) {
+                val supplier = EnvironmentTokenSupplier(BB_AUTH_USER_VAR, BB_AUTH_TOKEN_VAR, separator = ":")
+                supplier.token.shouldNotBeEmpty()
+                supplier.token.shouldContain(Regex("(.+):(.+)"))
```
```suggestion
                supplier.token.shouldContain(Regex(".+:.+"))
```
do you need to capture the groups?

DanySK commented on src/test/resources/SimpleTestClass.java:
```diff
@@ -0,0 +1,19 @@
+package org.example.javaparser.samples.Test;
+
+import java.util.Random;
+import java.util.stream.IntStream;
+
+/**
+ * This is a testing class.
+ */
+public class SimpleTestClass { // test class
+    public static void main(String[] args) {
+        var k = (new Random()).nextInt(1500);
+        IntStream.range(0, 1000).forEach( i -> {
+            k++;
+            if (!k%5) {
+                System.out.println(k + "5 is 0!");
+            }
+        });
+    }
+}
```
```suggestion
```

tassiLuca replied:
I use this Java class to test the tokenization process. Do you prefer I store this file in a different GH repo and then use the GHProvider to get it?

DanySK replied:
Oh, I see. It's fine as it is, then.

DanySK commented:
Nothing critical, enabling automerge.

tassiLuca commented on src/main/kotlin/org/danilopianini/plagiarismdetector/analyzer/technique/tokenization/java/JavaTokenizer.kt:
```diff
@@ -0,0 +1,59 @@
+                val result = tokens.toList()
+                tokens.clear()
+                return result
+            }
```
I changed the implementation in order to clear the list of tokens before each visit. Now should be more reasonable.
